frame/bags-list/src/list/mod.rs

```diff
@@ -384,6 +384,84 @@ impl<T: Config> List<T> {
 		})
 	}
 
+	/// Put `heavier_id` to the position directly in front of `lighter_id`. Both ids must be in
+	/// same bag  and the `weight_of` `lighter_id` must be less than that of `heavier_id`.
+	pub(crate) fn put_in_front_of(
+		lighter_id: &T::AccountId,
+		heavier_id: &T::AccountId,
+		weight_of: Box<dyn Fn(&T::AccountId) -> VoteWeight>,
+	) -> Result<(), crate::pallet::Error<T>> {
+		use crate::pallet;
+		use frame_support::ensure;
+
+		let lighter_node = Node::<T>::get(&lighter_id).ok_or(pallet::Error::IdNotFound)?;
+		let heavier_node = Node::<T>::get(&heavier_id).ok_or(pallet::Error::IdNotFound)?;
+
+		ensure!(lighter_node.bag_upper == heavier_node.bag_upper, pallet::Error::NotInSameBag);
+
+		// this is the most expensive check, so we do it last.
+		ensure!(weight_of(&heavier_id) > weight_of(&lighter_id), pallet::Error::NotHeavier);
+
+		// remove the heavier node from this list. Note that this removes the node from storage and
+		// decrements the node counter.
+		Self::remove(&heavier_id);
+
+		// re-fetch `lighter_node` from storage since it may have been updated when `heavier_node`
+		// was removed.
+		let lighter_node = Node::<T>::get(&lighter_id).ok_or(pallet::Error::IdNotFound)?;
+		Self::insert_at_unchecked(lighter_node, heavier_node);
+
+		// Increment the counter, which got decremented by `Self::remove`, since we just re-inserted
+		// `heavier_node`.
+		crate::CounterForListNodes::<T>::mutate(|prev_count| {
+			*prev_count = prev_count.saturating_add(1)
+		});
+
+		Ok(())
+	}
+
+	/// Insert `node` directly in front of `at`.
+	///
+	/// WARNINGS:
+	/// - we do not check that `node` or `at` are valid.
+	/// - this will panic if `at.bag_upper` is not a bag that already exists in storage.
+	/// - this does not update [`crate::CounterForListNodes`].
```

Comment:
why not update the counter here as well?

Reply:
flexibility - my thinking is that it is an internal function and updating the counter assumes it was decremented prior to inserting .. but thinking about it again this is called `insert*` so should probably update the counter because we assume this is "purely" an insert (as opposed to an update).

Reply:
fd1daeb0b0d6db4bc0f275f7ceea8ea308e6227d